### PR TITLE
Improve Trainer DataLoader Controls for Streaming and Multiprocessing

### LINKED ISSUE #34793: Allow to give the dataset multiprocessing_context
### Feature request

In Huggingface Trainer, allow to pass the multiprocessing context : https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader

### Motivation

For a dataset that is loaded on multiple cpu cores, sometimes the fork method creates problems (with polars for example) and the spawn method is more adapted. 

### Your contribution

I could do a PR. A fix could be to add one more parameter to Trainer and pass it to the Dataloader down the line.

### LINKED ISSUE #34867: Data prefetching does not occur for iterable datasets
### System Info

- `transformers` version: 4.46.1
- Platform: macOS-15.1-arm64-arm-64bit
- Python version: 3.11.10
- Huggingface_hub version: 0.26.2
- Safetensors version: 0.4.5
- Accelerate version: 1.0.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.5.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No

### Who can help?

@muellerzr @SunMarc 

### Reproduction

PR #28498 was meant to allow for specifying the Pytorch dataloader's `prefetch_factor` argument via the huggingface `dataloader_prefetch_factor` training argument. As we can see [on this line](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L989), this feature was added inside of a 
```python
if not isinstance(train_dataset, torch.utils.data.IterableDataset):
```
statement which results in prefetching never occurring for `IterableDataset`s (which seems like a mistake). There are a two other lines where the same error happens for [test](https://github.com/huggingface/transformers/blob/4e90b99ed916300b80bac9db793f2a96b2a87122/src/transformers/trainer.py#L1126) and [eval](https://github.com/huggingface/transformers/blob/4e90b99ed916300b80bac9db793f2a96b2a87122/src/transformers/trainer.py#L1084) dataloaders. Unless I'm missing something, I believe these lines can be moved out of the if condition and into other logic. 

### Expected behavior

Prefetching should work with iterable datasets.

### PR DESCRIPTION
**Pull request by muyihao**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47164)

# What does this PR do?

Follow-up to #43431. Adds three data-loading options to the `Trainer`:

1. **`dataloader_prefetch_factor` now applies to `IterableDataset`** — it was previously passed only in the map-style branch, so streaming datasets not work.
2. **New `dataloader_multiprocessing_context` arg** (`"fork"`/`"spawn"`/`"forkserver"`), forwarded to every DataLoader — lets users pick `"spawn"` for sources that are not fork-safe (e.g. HDFS via `pyarrow`) instead of the hard-coded start method.
3. **New `dataloader_in_order` arg** (default `True`) — forwards PyTorch's `in_order` flag (added in torch 2.6, pytorch/pytorch#141833) to reduce tail latency for shuffled `IterableDataset` streaming.

Fixes #34867

closes #34793 (EDIT from @qgallouedec)

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. -> Discussed in #43431; approved by @SunMarc on #34925.
- [x] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)? 
- [x] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

@SunMarc

### CODE REVIEW

**Comment by muyihao** [comment]
Hi @SunMarc 👋 — follow-up to #43431, continuing #34925.
Would appreciate a look when you get a chance 🙏

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -1736,6 +1767,27 @@ def _validate_args(self):
...
+
+        if not self.dataloader_in_order and is_torch_available():
+            from .pytorch_utils import is_torch_greater_or_equal_than_2_6
+
+            if not is_torch_greater_or_equal_than_2_6:
+                raise ValueError(
+                    "--dataloader_in_order=False requires PyTorch >= 2.6. Please upgrade PyTorch or leave"
+                    " --dataloader_in_order set to its default (True)."
+                )
+
```
Comment (qgallouedec) [inline src/transformers/training_args.py]: nit: I don't think that's very useful, a misconfiguration will fail early anyway.

**Review by qgallouedec — COMMENTED** [review]
This looks good to, me thanks, letting @SunMarc have the final word on this one

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -988,20 +989,28 @@ def _get_dataloader(
         # MPS requires forking if multiple workers are specified
         should_fork = torch.backends.mps.is_available() and self.args.dataloader_num_workers > 1
 
+        # A user-specified start method takes precedence over the MPS fork workaround.
+        multiprocessing_context = self.args.dataloader_multiprocessing_context
+        if multiprocessing_context is None and should_fork:
+            multiprocessing_context = "fork"
```
Comment (SunMarc) [inline src/transformers/trainer.py]: maybe we can put that in training_args.py instead of initializing it here ?

**Review by SunMarc — COMMENTED** [review]
Thanks ! Sorry for the delay ! Just a nit

File: src/transformers/trainer.py (hunk 2 above)
Comment (muyihao) [reply src/transformers/trainer.py]: agree, moved to training_args.py

File: src/transformers/training_args.py (hunk 1 above)
Comment (muyihao) [reply src/transformers/training_args.py]: removed

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -985,23 +986,23 @@ def _get_dataloader(
         else:
             data_collator = self._get_collator_with_removed_columns(self.data_collator, description=description)
 
-        # MPS requires forking if multiple workers are specified
-        should_fork = torch.backends.mps.is_available() and self.args.dataloader_num_workers > 1
-
```
Comment (SunMarc) [inline src/transformers/trainer.py:990]: i didn't mean to just remove that

File: src/transformers/trainer.py (hunk 2 above)
Comment (SunMarc) [reply src/transformers/trainer.py]: i don't see it

File: src/transformers/trainer.py (hunk 3 above)
Comment (muyihao) [reply src/transformers/trainer.py:990]: You're right, sorry. I misremembered macOS as defaulting to `fork`, so I removed this too aggressively. I've restored the MPS multi-worker `fork` default and moved the initialization into `TrainingArguments` as suggested.

Comment (SunMarc) [reply src/transformers/trainer.py:990]: thanks !

**Review by SunMarc — APPROVED** [review]
Thanks !
